service/history/historyCache.go

```diff
@@ -118,8 +118,11 @@ func (c *historyCache) getOrCreateWorkflowExecution(domainID string,
 	// This will create a closure on every request.
 	// Consider revisiting this if it causes too much GC activity
 	status := cacheNotReleased
-	releaseFunc := func() {
+	releaseFunc := func(err error) {
 		if atomic.CompareAndSwapInt32(&status, cacheNotReleased, cacheReleased) {
+			if err != nil {
```

Comment:
Let's create a workitem to review if we can skip clearing the cache for certain errors.  Like EntityNotFound.